crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -816,3 +541,321 @@ where
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+            }
+            None if tokens.contains_key(asset_type) => {
+                let (token, _denom, digit) = &tokens[asset_type];
+                let decoded_change = SignedAmount::from_masp_denominated(
+                    *val, *digit,
+                )
+                .map_err(|_| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+                let decoded_change =
+                    ValueSum::from_pair(token.clone(), decoded_change);
+                acc = acc.checked_sub(&decoded_change).ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in MASP value balance",
+                    ))
+                })?;
+            }
+            _ => {
+                let error =
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Unable to decode asset type",
+                    ));
+                tracing::debug!("{error}");
+                return Err(error);
+            }
+        }
+    }
+    if acc == ValueSum::from_sum(post.clone()) {
+        Ok(())
+    } else {
+        let error = Error::NativeVpError(native_vp::Error::SimpleMessage(
+            "MASP balance change not equal to Sapling value balance",
+        ));
+        tracing::debug!("{error}");
+        Err(error)
+    }
+}
+
+impl<'a, S, CA> NativeVp for MaspVp<'a, S, CA>
+where
+    S: StateRead,
+    CA: 'static + WasmCacheAccess,
+{
+    type Error = Error;
+
+    fn validate_tx(
+        &self,
+        tx_data: &Tx,
+        keys_changed: &BTreeSet<Key>,
+        _verifiers: &BTreeSet<Address>,
+    ) -> Result<()> {
+        let masp_keys_changed: Vec<&Key> =
+            keys_changed.iter().filter(|key| is_masp_key(key)).collect();
+
+        // Check that the transaction didn't write unallowed masp keys
+        if masp_keys_changed.iter().any(|key| {
+            !is_masp_transfer_key(key) && !is_masp_token_map_key(key)
+        }) {
+            return Err(Error::NativeVpError(native_vp::Error::SimpleMessage(
+                "Found modifications to non-allowed masp keys",
+            )));
+        }
+        // The token map can only be changed by a successful governance proposal
+        if masp_keys_changed
+            .iter()
+            .any(|key| is_masp_token_map_key(key))
+        {
+            self.is_valid_parameter_change(tx_data)?;
+        }
+        // The MASP transfer keys can only be changed by a valid Transaction
+        if masp_keys_changed
+            .iter()
+            .any(|key| is_masp_transfer_key(key))
+        {
+            self.is_valid_masp_transfer(tx_data, keys_changed)?;
```

Comment:
Do you think we should prevent normal masp transfers from governance proposals? It shouldn't be possible anyway given the way we construct the transaction to execute a proposal, but maybe for extra safety. If this became possible, since we don't emit events for governance transactions, the clients could not fetch this data leading to an impossibility to submit valid masp transfers

Reply:
Thanks, understood. Rectified in #3264 .